### PR TITLE
Enable pairwise DE in ingest pipeline (SCP-5766)

### PR DESCRIPTION
**Pull request by jlchang**
#### BACKGROUND & CHANGES

This update adds pairwise DE to ingest pipeline. A new `--differential-expression` ingest pipeline parameter, `--de-type` is instantiated as "rest" by default, allowing existing "one vs rest" invocations to remain unchanged. Pairwise invocations add two additional parameters `--group1` and `--group2`. ~~Scanpy's rank_genes_groups function will accept a list of annotation values for "--group1", while we won't be taking advantage of that capability, it does require `--group1` to be passed as a list~~. Both `--group1` and `--group2` ~~is~~ passed as ~~a simple~~ string.

#### MANUAL TESTING

Testing using ingest_pipeline CLI requires an AnnData file and cluster & metadata AnnData fragments extracted from the AnnData file. 
1. Download the following files (you may already have them from PR#372):
	**HTAPP-330-SMP-1082_compliant.h5ad**: gs://fc-2f8ef4c0-b7eb-44b1-96fe-a07f0ea9a982/test_Data/differential_expression/HTAPP-330-SMP-1082/HTAPP-330-SMP-1082_compliant.h5ad
	**HTAPP-330-SMP-1082_h5ad_frag.cluster.X_umap.tsv.gz**: gs://fc-2f8ef4c0-b7eb-44b1-96fe-a07f0ea9a982/test_Data/differential_expression/HTAPP-330-SMP-1082/HTAPP-330-SMP-1082_h5ad_frag.cluster.X_umap.tsv.gz (Note: file will lose the .gz suffix BUT will still need to be decompressed :(
	**HTAPP-330-SMP-1082_h5ad_frag.metadata.tsv.gz**: gs://fc-2f8ef4c0-b7eb-44b1-96fe-a07f0ea9a982/test_Data/differential_expression/HTAPP-330-SMP-1082/HTAPP-330-SMP-1082_h5ad_frag.metadata.tsv.gz (Note: file will lose the .gz suffix BUT will still need to be decompressed :(1. Initialize your dev environment and ```export BYPASS_MONGO_WRITES='yes'```
1. Run the following ingest pipeline job: 
`python ingest_pipeline.py --study-id addedfeed000000000000000 --study-file-id dec0dedfeed1111111111111 differential_expression --annotation-name leiden --de-type pairwise --group1 "1" --group2 "2" --annotation-type group --annotation-scope study --annotation-file HTAPP-330-SMP-1082_h5ad_frag.metadata.tsv.gz --cluster-file HTAPP-330-SMP-1082_h5ad_frag.cluster.X_umap.tsv.gz --cluster-name umap --matrix-file-path HTAPP-330-SMP-1082_compliant.h5ad  --matrix-file-type h5ad --study-accession SCPdev --differential-expression`
1. Confirm that the job generates "umap--leiden--1--2--study--wilcoxon.tsv"
1. [optional] Compare your output to the reference output file in the reference google bucket at gs://fc-2f8ef4c0-b7eb-44b1-96fe-a07f0ea9a982/test_Data/differential_expression/HTAPP-330-SMP-1082/umap--leiden--1--2--study--wilcoxon.tsv

### CODE REVIEW

File: ingest/de.py [hunk 1]
```diff
@@ -319,24 +331,63 @@ def all_match_ensembl_id_regex(adata, col_name):
...
+    def write_de_result(adata, group, annotation, rank_key, cluster_name, extra_params):
+        de_type =  extra_params.get("de_type")
+        method = extra_params.get("method")
+        annot_scope = extra_params.get("annotation_scope")
+        clean_annotation = DifferentialExpression.sanitize_string(annotation)
+        rank = sc.get.rank_genes_groups_df(adata, key=rank_key, group=group)
+        if DifferentialExpression.all_match_ensembl_id_regex(rank, "names"):
+            feature_name_rank = rank.merge(adata.var['feature_name'], left_on="names", right_index=True, how='left')
+            feature_name_rank.rename(columns={'names':'feature_id'}, inplace=True)
+            feature_name_rank.rename(columns={'feature_name':'names'}, inplace=True)
```
Comment (eweitz) [inline ingest/de.py:344]:
```suggestion
            feature_name_rank.rename(columns={'names': 'feature_id'}, inplace=True)
            feature_name_rank.rename(columns={'feature_name': 'names'}, inplace=True)
```

File: ingest/de.py [hunk 2]
```diff
@@ -319,24 +331,63 @@ def all_match_ensembl_id_regex(adata, col_name):
...
+        else:
+            msg = f'Unknown de_type, {de_type}'
+            print(msg)
+            log_exception(
+                DifferentialExpression.dev_logger, DifferentialExpression.de_logger, msg
+            )
+            raise ValueError(msg)
+        # Round numbers to 4 significant digits while respecting fixed point
+        # and scientific notation (note: trailing zeros are removed)
+        rank.to_csv(out_file, sep='\t', float_format='%.4g')
```
Comment (eweitz) [inline ingest/de.py:371]: Is there a scientific / engineering reason we're more precise for DE values (rounded to 4 significant digits) than for expression values and numeric annotation values ([rounded to 3](https://github.com/search?q=repo%3Abroadinstitute%2Fscp-ingest-pipeline+round&type=code))?

Either way, I see rounded DE values to 4 is pre-existing, so this seems OK in terms of this PR.

File: tests/test_de.py [hunk 3]
```diff
@@ -31,17 +31,32 @@ def get_annotation_labels(metadata, annotation, de_cells):
...
+        # rank_genes_groups accepts a list. For SCP pairwise, should be a list with one item
+        # converting list to string for incorporation into result filename
+        group1 = DifferentialExpression.sanitize_string(''.join(test_config["group1"]))
+        group2 = DifferentialExpression.sanitize_string(test_config["group2"])
+        expected_file = f'{sanitized_cluster_name}--{sanitized_annotation}--{group1}--{group2}--{annot_scope}--{method}.tsv'
+        print(f'expected_file, {expected_file}')
         assert os.path.exists(expected_file)
         found.append(expected_file)
-    return found
+        return found
```
Comment (eweitz) [inline tests/test_de.py:59]: Having `return found` after ending the `elif` block, and omitting `return found` in the `if` block, seems more conventional.

File: ingest/cli_parser.py [hunk 4]
```diff
@@ -274,6 +274,12 @@ def create_parser():
         help="Indicates that differential expression analysis should be invoked",
     )
 
+    parser_differential_expression.add_argument(
+        "--de-type",
+        default="rest",
+        help="Accepted values: 'pairwise' or 'rest'",
```
[the remark this answers is not in the document]
Comment (jlchang) [reply ingest/cli_parser.py]: Worth taking a minute to add `27d7be9`

File: ingest/cli_parser.py [hunk 5]
```diff
@@ -336,6 +342,18 @@ def create_parser():
         "--gene-file", help="Path to .genes.tsv file"
     )
 
+    # For pairwise analyses
+    parser_differential_expression.add_argument(
+        "--group1",
+        type=ast.literal_eval,
```
[the remark this answers is not in the document]
Comment (jlchang) [reply ingest/cli_parser.py:349]: I explained in PR description above
> Scanpy's rank_genes_groups function will accept a list of annotation values for "--group1", while we won't be taking advantage of that capability, it does require --group1 to be passed as a list. --group2 is passed as a simple string.
Will add comments to cli_parser to clarify for future us.

File: ingest/de.py [hunk 6]
```diff
@@ -78,8 +78,35 @@ def assess_annotation(annotation, metadata):
             )
             raise KeyError(msg)
         elif annotation_info == "group":
-            # DE annotations should be of TYPE group
-            return None
+            de_type =  extra_params.get("de_type")
```
[the remark this answers is not in the document]
Comment (jlchang) [reply ingest/de.py:81]: Hm, looks like my linter may not be working. I'll look into it.

File: ingest/de.py [hunk 7]
```diff
@@ -78,8 +78,35 @@ def assess_annotation(annotation, metadata):
...
+                        f'or group2, \"{group2}\", not found in metadata file'
+                        )
+                        log_exception(
+                            DifferentialExpression.dev_logger, DifferentialExpression.de_logger, msg
+                        )
+                        raise ValueError(msg)
+                else:
+                    msg = (
+                    f'Provided annotation value(s) group1, \"{group1}\", or '
+                    f'group2, \"{group2}\", were false-y'
```
[the remark this answers is not in the document]
Comment (jlchang) [reply ingest/de.py:100]: Thanks, I wasn't aware that one was more accepted than the other.

**Review by bistline — APPROVED** [review]

File: tests/test_de.py (hunk 3 above)
Comment (jlchang) [reply tests/test_de.py]: `e6287de`

File: ingest/de.py [hunk 8]
```diff
@@ -78,8 +78,35 @@ def assess_annotation(annotation, metadata):
             )
             raise KeyError(msg)
         elif annotation_info == "group":
-            # DE annotations should be of TYPE group
-            return None
+            de_type =  extra_params.get("de_type")
+            group1 = extra_params.get("group1")
+            group2 = extra_params.get("group2")
+            if  de_type == "pairwise":
```
[the remark this answers is not in the document]
Comment (jlchang) [reply ingest/de.py]: manually fixed in `640c1a4`

File: ingest/de.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (jlchang) [reply ingest/de.py]: manually fixed in `640c1a4`

File: ingest/de.py (hunk 1 above)
Comment (jlchang) [reply ingest/de.py]: manually fixed in `640c1a4`

File: ingest/de.py [hunk 9]
```diff
@@ -319,24 +331,63 @@ def all_match_ensembl_id_regex(adata, col_name):
...
+            new_column_order = ["names"] + [col for col in feature_name_rank.columns if col not in {"names", "feature_id"}] + ["feature_id"]
+            feature_name_rank = feature_name_rank[new_column_order]
+
+            rank = feature_name_rank
+        if DifferentialExpression.delimiter_in_gene_name(rank):
+            DifferentialExpression.extract_gene_id_for_out_file(rank)
+        if de_type == "rest":
+            clean_group = DifferentialExpression.sanitize_string(group)
+            out_file = f'{cluster_name}--{clean_annotation}--{clean_group}--{annot_scope}--{method}.tsv'
+            DifferentialExpression.de_logger.info(f"Writing DE output for {clean_group} vs restq")
```
[the remark this answers is not in the document]
Comment (jlchang) [reply ingest/de.py]: thanks for catching that typo! manually fixed in `640c1a4`

File: ingest/cli_parser.py [hunk 10]
```diff
@@ -336,6 +343,18 @@ def create_parser():
...
+        help="1st annotation group to use for pairwise DE analysis",
```
[the remark this answers is not in the document]
Comment (jlchang) [reply ingest/cli_parser.py]: `9ffe8c3`

File: ingest/cli_parser.py [hunk 11]
```diff
@@ -336,6 +343,18 @@ def create_parser():
...
+    )
+
+    parser_differential_expression.add_argument(
+        "--group2",
+        help="2nd annotation group to use for pairwise DE analysis",
```
[the remark this answers is not in the document]
Comment (jlchang) [reply ingest/cli_parser.py]: `9ffe8c3`

**Pull request merged**